docs/requirements/ADP-957.md

```diff
@@ -0,0 +1,156 @@
+# Basic definitions
+
+*   **Collateral inputs**, in a transaction, are used to cover transaction fees
+    in the case that a non-native script fails. The term **collateral** refers
+    to the total ada contained in the UTxOs referenced by collateral inputs.
+    Collateral inputs can **only** contain ada: they may **not** contain
+    non-ada assets.
+
+*   The protocol parameter `collateralPercent` is used to specify the
+    percentage of the total transaction fee its collateral must (at a minimum)
+    cover. The collateral inputs must not themselves be locked by a script.
+
+*   The protocol parameter `maxCollateralInputs` is used to limit,
+    additionally, the total number of collateral inputs, and thus the total
+    number of additional signatures that must be checked during validation.
+
+# Proposed requirements
+
+1.  The wallet, when requested to create a transaction requiring collateral,
+    should reserve one or more pure-ada UTxO entries to serve as collateral
+    inputs, where the total balance of collateral inputs is greater than or
+    equal to the minimum required, as determined by the `collateralPercent`
+    protocol parameter and the transaction fee.
+
+2.  The wallet, when requested to create a transaction requiring collateral,
+    should fail with an error if the total balance of pure-ada UTxO entries is
+    insufficient to cover the minimum collateral amount, as determined by the
+    `collateralPercent` protocol parameter and the transaction fee.
+
+3.  The wallet's create transaction operation should accept a parameter
+    `collateralPercentMaximum`, with a default value that can be overridden
+    by the caller. The wallet, when requested to create a transaction requiring
+    collateral, should fail with an error if it **is** possible to meet the
+    minimum amount, but **not** possible to find a set of pure-ada entries
+    whose total balance is less than or equal to
+    `collateralPercentMaximum × transactionFee`.
+
+4.  The wallet should ensure that any UTxO entries used for collateral are
+    **locked** until the transactions that include them are validated.
```

Comment:
Is locking needed as a requirement here? I mean, I guess, locking is something that we'd like to eventually have, but it is not necessarily part of req for collateral (Perhaps I'm wrong).

Reply:
I agree this requirement could be worded a bit better. I'll see if I can reword it.

My understanding: transactions that include collateral input will be marked as part of the pending set, as normal.

However, we'll have to be a bit careful here, as:
* for any given transaction `tx`:
  `inputs tx ∩ collateralInputs tx` can be non-empty.
* for any given pair of pending transactions `tx1` and `tx2`:
 `collateralInputs tx1 ∩ collateralInputs tx2` must be non-empty.

Reply:
@piotr-iohk I've now reworded the requirements to be:

  6.  The wallet application must ensure that UTxO entries used for collateral inputs in pending transactions are not available for reuse as collateral inputs in future transactions.  
      
  7.  The wallet application should consider a UTxO entry used for collateral in a pending transaction to be part of the UTxO set that is available for ordinary input selection in future transactions, unless that UTxO entry is also used as an ordinary input in a pending transaction.
  
Hopefully this is clearer, but let me know if you think anything is unclear.

Reply:
@piotr-iohk In light of our recent discussions, I've once again adjusted the wording, to read:

5.  The wallet application must ensure that **_UTxO entries used as collateral
    inputs_** in pending transactions are **not** available for reuse either as
    collateral inputs or as ordinary inputs in future transactions.

6.  The wallet application must ensure that **_UTxO entries used as ordinary
    inputs_** in pending transactions are **not** available for reuse either as
    collateral inputs or as ordinary inputs in future transactions.
    

7.  The wallet application, when requested to create a transaction requiring
    collateral, is permitted to select a set of collateral inputs that
    intersects with the set of ordinary inputs for that transaction:

        inputs t ≠ ∅ ∧ collateral_inputs t ≠ ∅

        =/=> (does not imply)

        inputs t     ∩ collateral_inputs t = ∅